### ISSUE TITLE
Text Snippets not available for CodeAction

### ISSUE THREAD

**Issue by NikolasKomonen** [microsoft/language-server-protocol#592]
From what I've seen it doesn't look like `CodeAction` supports snippets and defaults to plain text with no option to change it.

`CompletionItem`'s allow snippets through the `insertTextFormat` property.

If there is already a way could you please let me know.

Thanks.

**Comment by rcjsuen** [comment]
@NikolasKomonen Can you provide an example of the behaviour you are seeing? I'm not sure I understand your question/issue.

**Comment by angelozerr** [comment]
@NikolasKomonen post this question because in lsp4xml we have tried to use snippet in CodeAction for xml, see this issue https://github.com/angelozerr/lsp4xml/issues/185

**Comment by rcjsuen** [comment]
@NikolasKomonen Yeah, I don't think this is expected to work. I was able to reproduce your problem in the [Monaco Playground](https://microsoft.github.io/monaco-editor/playground.html).

```JavaScript
let uri = monaco.Uri.parse("uri://test.json");
let model = monaco.editor.createModel("xxx\nxxx\nxxx", "json", uri);

monaco.languages.registerCodeActionProvider('json', {
    provideCodeActions: function() {
        return [
            {
                title: "abc",
                edit: {
                    edits: [
                        {
                            resource: uri,
                            edits: [
                                {
                                    range: {
                                        startColumn: 1,
                                        startLineNumber: 1,
                                        endColumn: 1,
                                        endLineNumber: 1
                                    },
                                    text: "def($1)"
                                }
                            ]
                        }
                    ]
                }
            }
        ]
    }
});

monaco.editor.create(document.getElementById("container"), {
    model: model,
    language: "json",
    lightbulb: {
        enabled: true
    }
});
```

**Comment by rcjsuen** [comment]
Oh, wow. I totally thought this was an issue in the [Microsoft/monaco-editor](https://github.com/Microsoft/monaco-editor/) repository! :(

In any case, I agree with your assessment that there is no way for snippets to be used in a `CodeAction`.

**Comment by angelozerr** [comment]
Thanks @rcjsuen for your feedback, it should be cool if LSP could specify this feature `snippet support for CodeAction`.

**Comment by NikolasKomonen** [comment]
@rcjsuen Thanks for the feedback

**Comment by heejaechang** [comment]
@dbaeumer any plan to support this? are you open to this feature request such as accepting contribution?

**Comment by dbaeumer** [comment]
@heejaechang defining this in the spec is the easy part. We need a client implementation as well and at least VS Code has no support for this right now.

**Comment by matklad** [comment]
FWIW, in rust-analyzer we implement this already via a protocol extension and a hacky vscode impl:

* [spec](https://github.com/rust-analyzer/rust-analyzer/blob/911ff38eae53eeba7e94a6b01d44eef568b8c476/docs/dev/lsp-extensions.md#snippet-textedit)
* [impl](https://github.com/rust-analyzer/rust-analyzer/blob/911ff38eae53eeba7e94a6b01d44eef568b8c476/editors/code/src/snippets.ts#L5-L21)

Some other clients implement this extension as well, for example, [lsp-mode](https://github.com/emacs-lsp/lsp-mode/blob/64412b8c530163db7955730f1853a8ef8b3a203c/clients/lsp-rust.el#L664)

**Comment by heejaechang** [comment]
@matklad thank you, I will take a look! @dbaeumer even though it is not supported by any client, would be nice at least in spec, so some client can support it.

**Comment by dbaeumer** [comment]
@heejaechang we decided that we only add something to the spec if we have an implementation in a client. Otherwise things are hard to spec and to get feedback on.

**Comment by michaelmesser** [comment]
I'm confused. Is lsp-mode not a client?

**Comment by KamasamaK** [comment]
@dbaeumer said
> @heejaechang defining this in the spec is the easy part. We need a client implementation as well and at least VS Code has no support for this right now.

VS Code now supports this since v1.72. See microsoft/vscode#145374

**Comment by dbaeumer** [comment]
Something to work on for 3.18.

**Comment by heejaechang** [comment]
@dbaeumer can you make sure SnippetTextEdit is allowed all places where LSP returns TextEdit not just workspace edits? things like "FormatOnType" which returns TextEdit[].

thank you!

**Comment by rchiodo** [comment]
If this was supported for all TextEdits, it would make it possible for us to implement this feature for pylance:
https://github.com/microsoft/pylance-release/discussions/2615

Without it, the cursor comes out weird.

**Comment by heejaechang** [comment]
@jrieken how about other places where TextEdit is used in vscode api such as `provideOnTypeFormattingEdits`, `provideDocumentRangeFormattingEdits` and more? shouldn't all they support SnippetTextEdit?

**Comment by jrieken** [comment]
@heejaechang not yet decided. Your case is valid but generally full blown snippets while users type could be asking for trouble. E.g snippets can force you into an interactive mode, e.g have you to select a choice element, have multiple placeholders, etc pp. Having this happen while I type along (or on save) doesn't seem like a good idea. 

> ... this feature for pylance: https://github.com/microsoft/pylance-release/discussions/2615

Can't you do something like @meganrogge has done with https://marketplace.visualstudio.com/items?itemName=meganrogge.template-string-converter. AFAIK this isn't a format-on-type provider but just a typing listener

**Comment by rchiodo** [comment]
@jrieken thanks for the suggestion. We were holding Megan's solution as a last resort because it requires a lot of work in the extension (whereas the language server already has this information). So, the hope was to be able to just change selection as part of the edits returned from the `provideDocumentRangeFormattingEdits`.

**Comment by jrieken** [comment]
> because it requires a lot of work in the extension (whereas the language server already has this information)

Understood, I would still implement the logic in the language server but with a custom route/message that the extension invokes when certain things happen.

**Comment by deakjahn** [comment]
Another use case for them is refactoring code actions. Say, I have a code action that converts a `foreach` loop into a `for` loop. During refactoring, I introduce a new index variable that I'd need to allow the user to rename in snippet edit mode. Yes, I do it now by sending it back to the client and doing it from there. Possible but not really elegant, snippets should be supported everywhere.

**Comment by puremourning** [comment]
Doesn’t need to be a snippet. Could be an argument (like rename). Snippets are not a panacea of ux

**Comment by dbaeumer** [comment]
This got implemented for 3.18 be having snippet support in WorkspaceEdits
